src/frontend/mod.rs

```diff
@@ -62,212 +49,417 @@ impl<'a, B: Backend<'a>> Frontend<'a, B> {
+        }
+        String::from_utf8(out).expect("invalid utf8")
+
     }
 
-    fn handle_inline_code(&mut self) -> Event<'a> {
-        // peek if code is math mode
-        let inner = self.parser.next().unwrap();
-        let text = match inner {
+    fn convert_inline_code(&mut self) {
+        // check if code is math mode
+        let mut text = match self.parser.next().unwrap() {
             CmarkEvent::Text(text) => text,
-            e => unreachable!("InlineCode should always be followed by Text but was fallowed by {:?}", e),
+            CmarkEvent::End(CmarkTag::Code) => {
+                self.buffer.push_back(Event::Start(Tag::InlineCode));
+                self.buffer.push_back(Event::End(Tag::InlineCode));
+                return;
+            }
+            e => unreachable!("InlineCode should always be followed by Text or End(Code) but was fallowed by {:?}", e),
         };
-        if text.starts_with("$ ") {
-            let text = match text {
-                Cow::Borrowed(s) => Cow::Borrowed(&s[2..]),
-                Cow::Owned(mut s) => {
-                    s.drain(..2);
-                    Cow::Owned(s)
-                }
-            };
-            self.buffer.push_back(Event::Text(text));
-            self.state = State::Math;
-            Event::Start(Tag::InlineMath)
+        let tag = if text.starts_with("$ ") {
+            text.truncate_left(2);
+            Tag::InlineMath
         } else {
-            self.buffer.push_back(Event::Text(text));
-            Event::Start(Tag::InlineCode)
-        }
+            Tag::InlineCode
+        };
+        self.buffer.push_back(Event::Start(tag.clone()));
+        self.buffer.push_back(Event::Text(text));
+        self.convert_until_end_inclusive(|t| if let CmarkTag::Code = t { true } else { false });
+        self.buffer.push_back(Event::End(tag));
     }
 
-    fn handle_code_block(&mut self, lang: Cow<'a, str>) -> Event<'a> {
+    fn convert_code_block(&mut self, lang: Cow<'a, str>, mut cskvp: Option<Cskvp<'a>>) {
         let lang = match lang {
             Cow::Borrowed(s) => s,
-            Cow::Owned(_) => unreachable!(),
+            Cow::Owned(_) => unreachable!("CodeBlock language should be borrowed"),
         };
-        let (single, mut double) = parse_attributes(lang);
-        let res = match single[0] {
-            "equation" | "math" | "$$" => {
-                self.state = State::Equation;
-                Event::Start(Tag::Equation)
+
+        // check if language has label/config
+        let language;
+        if let Some(pos) = lang.find(',') {
+            language = &lang[..pos];
+            if cskvp.is_some() {
+                // TODO: error
+                println!("Code has both prefix and inline style labels / config, ignoring both");
+                // don't print warnings about unused properties
+                // will be cleaned up as it's on the stack anyways
+                mem::forget(cskvp.take());
```

Comment:
I don't fully understand what you mean here. What would `drain` write into? My idea was that I *don't* need to call a function to print the warning about possibly unused attributes everywhere. I do agree that `mem::forget` here seems weird, so maybe I should add a method which consumes everything and thus not printing a message. But then again `mem::forget` does pretty much the same.

Reply:
> so maybe I should add a method which consumes everything and thus not printing a message.

I meant pretty much the same, I think:

```
impl<'a> Cskvp<'a> {
    fn drain(&mut self, write: impl Write) -> io::Result<()> {
        if let Some(label) = self.label.take() {
            writeln!(o, "label ignored: {}", label)?;
        }
        if let Some(figure) = self.figure.take() {
            writeln!(o, "figure config ignored: {}", figure)?;
        }
        if let Some(caption) = self.caption.take() {
            writeln!(o, "caption ignored: {}", caption)?;
        }
        for (k, v) in self.double.drain() { // here we already have drain!
            writeln!(o, "Unknown attribute `{}={}`", k, v)?;
        }
        for attr in self.single.drain(..) {
            writeln!(o, "Unknown attribute `{}`", attr)?;
        }
        Ok(())
    }
}

impl<'a> Drop for Cskvp<'a> {
    fn drop(&mut self) {
        self.drain(stdin().lock()).unwrap();
    }
}
```

with usage:

```
// replaces mem::forget(), to just clear any unused parts.
cskvp.drain(io::sink()).unwrap()
```

Improvement: `mem::forget` will currently leak the memory of all `Cow::Owned` and the hashmap for example.

Reply:
> `mem::forget` will currently leak the memory of the hashmap and the vec, and could potentially leak `Cow::Owned` when parsing gets improved to handle quote escapes.

Ouch, totally forgot about that. That's no good. Definitely requires a change!

Reply:
Unfortunately that solution doesn't really work once we use the `log` crate, because we'd like to have the `warn!` output's prefix in each printed line as well. So the `drain` method should just `take()` all options and `clear()` the Vec and HashMap.

Reply:
Ah yes. Alternatively, could specify a new trait for the `drain` operation, implemented for `T: io::Write` but in the future also for a private `struct Warn` that gets called upon in `Drop`. That may be convoluted though.

Reply:
I ended up writing a `.clear()` method and am calling that to reset the `cskvp` to its `Default` state.